pkg/piecestore/psserver/readerwriter.go

```diff
@@ -77,5 +87,18 @@ func NewStreamReader(s *Server, stream pb.PieceStoreRoutes_StoreServer) *StreamR
 
 // Read -- Read method for piece download from stream
 func (s *StreamReader) Read(b []byte) (int, error) {
-	return s.src.Read(b)
+	if s.sofar >= s.bandwidthRemaining {
+		return 0, StreamWriterError.New("out of bandwidth")
+	}
+	if s.sofar >= s.spaceRemaining {
+		return 0, StreamWriterError.New("out of space")
+	}
+
+	n, err := s.src.Read(b)
+	s.sofar += int64(n)
```

Comment:
Why don't we have the checks if space and bandwidth were exceeded here?

For example, imagine we have 1 KB space left before calling `StreamReader.Read`. Calling `s.src.Read(b)` reads 32 KB. Now we would have 31 KB more data than space available and we don't return an error. The error will be returned on the next `StreamReader.Read`, which is late.

Reply:
We should also have a test case covering the above use case.

Reply:
I didn't mean to use another Read method, but to do the checks from the beginning of this method right after the calling `s.src.Read` too.

Well, maybe it does not make sense to check if the available bandwidth is depleted, because the excess bandwidth of calling `s.src.Read` is already wasted. But perhaps it makes sense to check if the available space is depleted and fail early here. This way we will prevent the attempt of writing the excess data to the local storage.

Reply:
I was actually originally doing the check here. I guess it makes sense to fail as soon as we know instead of on the next iteration. I will make the change and add a test case

Reply:
It still make sense to have them at the beginning of the method to prevent calling the underlying reader. And as I stated it only make sense to the the space check right after calling the underlying reader.